// @workflow/world-vercel/dist/utils.js
import { getVercelOidcToken } from '@vercel/oidc';
import { WorkflowAPIError } from '@workflow/errors';
import { ZodError } from 'zod';
export const DEFAULT_RESOLVE_DATA_OPTION = 'all';
export function dateToStringReplacer(_key, value) {
    if (value instanceof Date) {
        return value.toISOString();
    }
    return value;
}
export const getHttpUrl = (config) => {
    const projectConfig = config?.projectConfig;
    const defaultUrl = 'https://vercel-workflow.com/api';
    const defaultProxyUrl = 'https://api.vercel.com/v1/workflow';
    const usingProxy = Boolean(config?.baseUrl || (projectConfig?.projectId && projectConfig?.teamId));
    const baseUrl = config?.baseUrl || (usingProxy ? defaultProxyUrl : defaultUrl);
    return { baseUrl, usingProxy };
};
export const getHeaders = (config) => {
    const projectConfig = config?.projectConfig;
    const headers = new Headers(config?.headers);
    if (projectConfig) {
        headers.set('x-vercel-environment', projectConfig.environment || 'production');
        if (projectConfig.projectId) {
            headers.set('x-vercel-project-id', projectConfig.projectId);
        }
        if (projectConfig.teamId) {
            headers.set('x-vercel-team-id', projectConfig.teamId);
        }
    }
    return headers;
};
export async function getHttpConfig(config) {
    const headers = getHeaders(config);
    const token = config?.token ?? (await getVercelOidcToken());
    if (token) {
        headers.set('Authorization', `Bearer ${token}`);
    }
    const { baseUrl, usingProxy } = getHttpUrl(config);
    return { baseUrl, headers, usingProxy };
}
export async function makeRequest({ endpoint, options = {}, config = {}, schema, }) {
    const { baseUrl, headers } = await getHttpConfig(config);
    headers.set('Content-Type', 'application/json');
    const url = `${baseUrl}${endpoint}`;
    const response = await fetch(url, {
        ...options,
        headers,
    });
    if (!response.ok) {
        const errorData = (await response.json().catch(() => ({})));
        if (process.env.DEBUG === '1') {
            const stringifiedHeaders = Array.from(headers.entries())
                .map(([key, value]) => `-H "${key}: ${value}"`)
                .join(' ');
            console.error(`Failed to fetch, reproduce with:\ncurl -X ${options.method} ${stringifiedHeaders} "${url}"`);
        }
        throw new WorkflowAPIError(errorData.message ||
            `${options.method ?? 'GET'} ${endpoint} -> HTTP ${response.status}: ${response.statusText}`, { url, status: response.status, code: errorData.code });
    }
    try {
        const text = await response.text();
        return schema.parse(JSON.parse(text));
    }
    catch (error) {
        if (error instanceof ZodError) {
            throw new WorkflowAPIError(`Failed to parse server response for ${options.method ?? 'GET'} ${endpoint}: ${error.message}`, { url, cause: error });
        }
        throw new WorkflowAPIError(`Failed to parse server response for ${options.method ?? 'GET'} ${endpoint}`, { url, cause: error });
    }
}
